3rdparty/ndsrvp/src/warpPerspective.cpp

```diff
@@ -0,0 +1,150 @@
+
+                            int32x2_t vX = {saturate_cast<int>(fX1), saturate_cast<int>(fX2)};
+                            int32x2_t vY = {saturate_cast<int>(fY1), saturate_cast<int>(fY2)};
+
+                            int32x2_t vx = __nds__v_sclip32(__nds__v_sra32(vX, INTER_BITS), 15);
+                            int32x2_t vy = __nds__v_sclip32(__nds__v_sra32(vY, INTER_BITS), 15);
+
+                            *(uint16x4_t*)(xy + x1 * 2) = (uint16x4_t)__nds__pkbb16((unsigned long)vy, (unsigned long)vx);
+
+                            uint32x2_t valpha = __nds__v_uadd32(__nds__v_sll32((uint32x2_t)(vY & vmask), INTER_BITS), (uint32x2_t)(vX & vmask));
+                            *(int16x2_t*)(alpha + x1) = (int16x2_t) { (short)(valpha[0]), (short)(valpha[1]) };
+                        }
+
+                        for (; x1 < bw; x1++) {
+                            double W = W0 + M[6] * x1;
+                            W = W ? INTER_TAB_SIZE / W : 0;
+                            double fX = std::max((double)INT_MIN, std::min((double)INT_MAX, (X0 + M[0] * x1) * W));
+                            double fY = std::max((double)INT_MIN, std::min((double)INT_MAX, (Y0 + M[3] * x1) * W));
+                            int X = saturate_cast<int>(fX);
+                            int Y = saturate_cast<int>(fY);
+
+                            xy[x1 * 2] = saturate_cast<short>(X >> INTER_BITS);
+                            xy[x1 * 2 + 1] = saturate_cast<short>(Y >> INTER_BITS);
+                            alpha[x1] = (short)((Y & (INTER_TAB_SIZE - 1)) * INTER_TAB_SIZE + (X & (INTER_TAB_SIZE - 1)));
+                        }
+                    }
+                }
+
+                if (interpolation == CV_HAL_INTER_NEAREST)
+                    remap(src, dpart, _XY, Mat(), interpolation, borderType, borderValue);
+                else {
+                    Mat _matA(bh, bw, CV_16U, A);
+                    remap(src, dpart, _XY, _matA, interpolation, borderType, borderValue);
+                }
+            }
+        }
+    }
+
+private:
+    Mat src;
+    Mat dst;
+    const double* M;
+    int interpolation, borderType;
+    Scalar borderValue;
+};
+
+int ndsrvp_warpPerspective(int src_type,
+    const uchar* src_data, size_t src_step, int src_width, int src_height,
+    uchar* dst_data, size_t dst_step, int dst_width, int dst_height,
+    const double M[9], int interpolation, int borderType, const double borderValue[4])
+{
+    Mat src(Size(src_width, src_height), src_type, const_cast<uchar*>(src_data), src_step);
+    Mat dst(Size(dst_width, dst_height), src_type, dst_data, dst_step);
+
+    Range range(0, dst.rows);
+    NdsrvpWarpPerspectiveInvoker invoker(src, dst, M, interpolation, borderType, Scalar(borderValue[0], borderValue[1], borderValue[2], borderValue[3]));
+    parallel_for_(range, invoker, dst.total() / (double)(1 << 16));
+    return CV_HAL_ERROR_OK;
+}
+
+}
```

Comment:
Please add single empty line at the end of all files.

Reply:
Done.